<a href="https://colab.research.google.com/github/Databyrk/loan-approval-prediction/blob/main/Loan_Aprovel_prediction_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [59]:
from google.colab import files
uploaded=files.upload()
print(uploaded)

Saving loan_approval_1000.csv to loan_approval_1000.csv
{'loan_approval_1000.csv': b'Age,Income,Loan_Amount,Credit_Score,Employment_Type,Loan_Approved\n59,77761.0,417505,337,Salaried,0\n49,141178.0,328226,487,Salaried,0\n35,56368.0,312805,356,Self-Employed,0\n28,26783.0,218484,319,Salaried,0\n41,131530.0,315064,572,Self-Employed,0\n59,61914.0,295378,418,Self-Employed,0\n39,90507.0,391705,882,Salaried,1\n43,147738.0,321090,815,Self-Employed,1\n31,99026.0,275314,520,Self-Employed,0\n31,135893.0,488716,825,Salaried,1\n44,131255.0,285612,814,Salaried,1\n56,130616.0,328453,635,Salaried,0\n44,87526.0,105230,354,Salaried,0\n23,90835.0,423966,799,Unemployed,1\n42,65765.0,473243,835,,0\n22,39956.0,424484,461,Self-Employed,0\n44,107895.0,109321,893,Salaried,1\n50,66240.0,182028,458,,0\n58,141008.0,281447,487,Unemployed,0\n22,57148.0,386275,359,Self-Employed,0\n41,114270.0,491124,618,Salaried,0\n53,147978.0,338107,510,Salaried,0\n32,24757.0,151757,705,,0\n42,39097.0,482659,408,Salaried,0\n45,3039

## **Import Required Libraries**

In [77]:
import pandas as pd
import numpy as np

## **Load the Dataset**

In [62]:
df=pd.read_csv("loan_approval_1000.csv")

In [63]:
df

,Age,Income,Loan_Amount,Credit_Score,Employment_Type,Loan_Approved
0,59,77761.0,417505,337,Salaried,0
1,49,141178.0,328226,487,Salaried,0
2,35,56368.0,312805,356,Self-Employed,0
3,28,26783.0,218484,319,Salaried,0
4,41,131530.0,315064,572,Self-Employed,0
...,...,...,...,...,...,...
995,58,112447.0,259717,572,Unemployed,0
996,22,134994.0,314167,544,Unemployed,0
997,27,109899.0,169654,749,Salaried,1
998,38,40049.0,240868,546,Unemployed,0


## **Handle Missing Values**

In [64]:
df.isna().sum()

,0
Age,0
Income,50
Loan_Amount,0
Credit_Score,0
Employment_Type,50
Loan_Approved,0


In [65]:
len(df)

1000

In [100]:
df["Employment_Type"].value_counts()

,count
Employment_Type,
Salaried,471
Self-Employed,287
Unemployed,192


In [99]:
mu = df["Income"].mean()
mod = df["Employment_Type"].mode()[0]

In [102]:
mu


np.float64(84881.75368421053)

In [103]:
mod

'Salaried'

In [104]:
df["Employment_Type"].mode()

,Employment_Type
0,Salaried


In [119]:
df["Income"]=df["Income"].fillna(mu)
df["Employment_Type"]=df["Employment_Type"].fillna(mod)

In [109]:
df.isna().sum()

,0
Age,0
Income,0
Loan_Amount,0
Credit_Score,0
Employment_Type,0
Loan_Approved,0


##**Encode Categorical Variables & Data Preprocessing**

In [120]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

df["Employment_Type"] = le.fit_transform(df["Employment_Type"])
df

,Age,Income,Loan_Amount,Credit_Score,Employment_Type,Loan_Approved
0,59,77761.0,417505,337,0,0
1,49,141178.0,328226,487,0,0
2,35,56368.0,312805,356,1,0
3,28,26783.0,218484,319,0,0
4,41,131530.0,315064,572,1,0
...,...,...,...,...,...,...
995,58,112447.0,259717,572,2,0
996,22,134994.0,314167,544,2,0
997,27,109899.0,169654,749,0,1
998,38,40049.0,240868,546,2,0


In [122]:
le.classes_

array([0, 1, 2])

##**Split Data into Training and Testing Sets**

In [123]:
x = df.drop("Loan_Approved",axis=1)
y = df["Loan_Approved"]

from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2)



##**Build the Machine Learning Model**
##**Train the Logistic Regression Model**

In [142]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(max_iter=1000)
lr.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

##**Model Prediction**

In [143]:
y_pred =lr.predict(x_test)

In [130]:
y_pred

array([0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0])

In [144]:
y_test

,Loan_Approved
453,0
513,0
310,0
116,1
765,0
...,...
210,0
562,0
192,0
92,1


##**Model Evaluation**

In [159]:
from sklearn.metrics import classification_report

print (classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.81      0.93      0.87       132
           1       0.82      0.59      0.68        68

    accuracy                           0.81       200
   macro avg       0.82      0.76      0.78       200
weighted avg       0.82      0.81      0.81       200



##** Hyperparameter Tuning**

In [155]:
lr = LogisticRegression(max_iter=1000)
lr.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

In [157]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.815

##**Final Results**

The Loan Approval Prediction project successfully developed a Machine Learning model to predict whether a loan application is likely to be approved or rejected.

The project involved data cleaning, exploratory data analysis, preprocessing, categorical encoding, feature scaling, model training, and hyperparameter tuning.

##Model Performance

*  Machine Learning Algorithm: Logistic Regression
*  Training Data: 80% of the dataset
*  Testing Data: 20% of the dataset
*  Accuracy:81.5%
*  Precision: for class 0 81% & class 1 82%
*  Recall: class 0 93% & class 1 59%
*  F1-Score:  87% &  68%

Hyperparameter tuning was performed to improve the performance of the Logistic Regression model and identify suitable model parameters.

The final model was evaluated using accuracy, confusion matrix, precision, recall, and F1-score to understand its predictive performance.




## Conclusion

This project successfully developed a **Loan Approval Prediction model using Logistic Regression**. The dataset was processed through data cleaning, preprocessing, feature transformation, and model training before evaluating its performance on unseen test data.

The model achieved an overall **accuracy of 81.5%**. It achieved **81% precision, 93% recall, and 87% F1-score for Class 0**, while for **Class 1**, it achieved **82% precision, 59% recall, and 68% F1-score**.

The results show that the model performs well overall, particularly in identifying Class 0 cases. However, the lower recall for Class 1 indicates that some Class 1 cases were not correctly identified. This suggests that further improvements such as feature engineering, class balancing, and additional hyperparameter tuning could help improve the model's ability to identify both classes.

Overall, this project demonstrates an end-to-end Machine Learning workflow for predicting loan approval and provided practical experience in **data preprocessing, exploratory data analysis, Logistic Regression, model evaluation, and hyperparameter tuning**.
